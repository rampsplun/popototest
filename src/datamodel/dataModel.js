var dataModel = {};

dataModel.idGen = 0;

dataModel.generateId = function () {
    return dataModel.idGen++;
};

dataModel.nodes = [];
dataModel.links = [];

dataModel.getRootNode = function () {
    return dataModel.nodes[0];
};

export default dataModel