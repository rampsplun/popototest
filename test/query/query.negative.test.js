import provider from "../../src/provider/provider";
import query from '../../src/query/query.js'
import dataModel from "../../src/datamodel/dataModel";

describe("one branch", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node",
                type: 1,
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("WHERE (NOT exists (path-to-branch))", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});

describe("one branch with one value", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node",
                type: 1,
                value: [{label: "Node", attributes: {Nid: "Vid", Nname: "Vname"}}],
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("WHERE (NOT exists (path-to-branch-with-value))", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});

describe("one branch with two value", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node",
                type: 1,
                value: [
                    {label: "Node", attributes: {Nid: "Vid", Nname: "Vname"}},
                    {label: "Node", attributes: {Nid: "Vid2", Nname: "Vname2"}}
                ],
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("WHERE (NOT exists (path-to-branch-with-value))", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});

describe("one long-branch", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node1",
                type: 1,
            },
            {
                id: 3,
                label: "Node",
                internalLabel: "node2",
                type: 1,
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO2",
                source: dataModel.nodes[1],
                target: dataModel.nodes[2],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("WHERE (NOT exists (path-to-long-branch))", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});

describe("one long-branch with value leaf", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node1",
                type: 1,
            },
            {
                id: 3,
                label: "Node",
                internalLabel: "node2",
                value: [{label: "Node", attributes: {Nid: "Vid", Nname: "Vname"}}],
                type: 1,
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO2",
                source: dataModel.nodes[1],
                target: dataModel.nodes[2],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("WHERE (NOT exists (path-to-long-branch-with-value))", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});

describe("one long-branch with value mid", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node1",
                value: [{label: "Node", attributes: {Nid: "Vid", Nname: "Vname"}}],
                type: 1,
            },
            {
                id: 3,
                label: "Node",
                internalLabel: "node2",
                type: 1,
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO2",
                source: dataModel.nodes[1],
                target: dataModel.nodes[2],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("WHERE (NOT exists (path-to-long-branch-with-value))", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});

describe("one long-branch with value mid and leaf", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node1",
                value: [{label: "Node", attributes: {Nid: "Vid", Nname: "Vname"}}],
                type: 1,
            },
            {
                id: 2,
                label: "Node",
                internalLabel: "node2",
                value: [{label: "Node", attributes: {Nid: "Vid2", Nname: "Vname2"}}],
                type: 1,
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO2",
                source: dataModel.nodes[1],
                target: dataModel.nodes[2],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("WHERE (NOT exists (path-to-long-branch-with-value))", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});

describe("one long-branch with NOT value mid", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node1",
                value: [{label: "Node", attributes: {Nid: "Vid", Nname: "Vname"}}],
                isNegative: true,
                type: 1,
            },
            {
                id: 3,
                label: "Node",
                internalLabel: "node2",
                type: 1,
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO2",
                source: dataModel.nodes[1],
                target: dataModel.nodes[2],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("WHERE (NOT exists (path-to-long-branch-with-value))", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});

describe("one very long-branch with value leaf and reverse order", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Label1",
                internalLabel: "label1",
                isNegative: false,
                type: 1,
            },
            {
                id: 2,
                label: "Label2",
                isParentRelReverse: true,
                internalLabel: "label2",
                type: 1,
            },
            {
                id: 3,
                label: "Label3",
                internalLabel: "label3",
                value: [{label: "Label3", attributes: {Nid: "Vid3", Nname: "Vname3"}}],
                type: 1,
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO1",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO2",
                source: dataModel.nodes[1],
                target: dataModel.nodes[2],
                type: 1
            },
            {
                id: 2,
                label: "LINKED_TO3",
                source: dataModel.nodes[2],
                target: dataModel.nodes[3],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Label1": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            },
            "Label2": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            },
            "Label3": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("generation", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});


describe("one branch generateNegativeNodeValueConstraints", function () {
    beforeEach(() => {
        dataModel.nodes = [
            {
                id: 0,
                label: "Root",
                internalLabel: "root",
                type: 0
            },
            {
                id: 1,
                label: "Node",
                internalLabel: "node",
                type: 1,
                isNegative: true
            }
        ];

        dataModel.links = [
            {
                id: 1,
                label: "LINKED_TO",
                source: dataModel.nodes[0],
                target: dataModel.nodes[1],
                type: 1
            }
        ];

        provider.node.Provider = {
            "Root": {
                returnAttributes: ["name", "id"],
                constraintAttribute: "id",
            },
            "Node": {
                returnAttributes: ["Nname", "Nid"],
                constraintAttribute: "Nid",
                generateNegativeNodeValueConstraints: function (node) {
                    return {
                        whereElements: [
                            "(NOT exists (root)-[]->())"
                        ],
                        parameters: {
                            customParam: "Custom Param Value"
                        }
                    }
                },
            }
        };
    });

    afterEach(() => {
        delete provider.node.Provider;
    });

    test("generateResultQuery", () => {
        var generateResultQuery = query.generateResultQuery(false);
        expect(generateResultQuery.statement).toMatchSnapshot();
        expect(generateResultQuery.parameters).toMatchSnapshot();
    });
});
