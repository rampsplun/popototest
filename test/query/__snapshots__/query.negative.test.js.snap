// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one branch WHERE (NOT exists (path-to-branch)) 1`] = `"MATCH (root:\`Root\`) WHERE (NOT exists((root)-[:\`LINKED_TO\`]->(:\`Node\`))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one branch WHERE (NOT exists (path-to-branch)) 2`] = `
Object {
  "limit": 100,
}
`;

exports[`one branch generateNegativeNodeValueConstraints generateResultQuery 1`] = `"MATCH (root:\`Root\`) WHERE (NOT exists (root)-[]->()) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one branch generateNegativeNodeValueConstraints generateResultQuery 2`] = `
Object {
  "customParam": "Custom Param Value",
  "limit": 100,
}
`;

exports[`one branch with one value WHERE (NOT exists (path-to-branch-with-value)) 1`] = `"MATCH (root:\`Root\`) WHERE (NOT exists((root)-[:\`LINKED_TO\`]->(:\`Node\`{Nid:$node_Nid}))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one branch with one value WHERE (NOT exists (path-to-branch-with-value)) 2`] = `
Object {
  "limit": 100,
  "node_Nid": "Vid",
}
`;

exports[`one branch with two value WHERE (NOT exists (path-to-branch-with-value)) 1`] = `"MATCH (root:\`Root\`) WHERE (NOT exists((root)-[:\`LINKED_TO\`]->(:\`Node\`{Nid:$node_Nid_0}))) AND (NOT exists((root)-[:\`LINKED_TO\`]->(:\`Node\`{Nid:$node_Nid_1}))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one branch with two value WHERE (NOT exists (path-to-branch-with-value)) 2`] = `
Object {
  "limit": 100,
  "node_Nid_0": "Vid",
  "node_Nid_1": "Vid2",
}
`;

exports[`one long-branch WHERE (NOT exists (path-to-long-branch)) 1`] = `"MATCH (root:\`Root\`) WHERE (NOT exists((root)-[:\`LINKED_TO1\`]->(:\`Node\`)-[:\`LINKED_TO2\`]->(:\`Node\`))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one long-branch WHERE (NOT exists (path-to-long-branch)) 2`] = `
Object {
  "limit": 100,
}
`;

exports[`one long-branch with NOT value mid WHERE (NOT exists (path-to-long-branch-with-value)) 1`] = `"MATCH (root:\`Root\`) WHERE (NOT exists((root)-[:\`LINKED_TO1\`]->(:\`Node\`{Nid:$node1_Nid}))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one long-branch with NOT value mid WHERE (NOT exists (path-to-long-branch-with-value)) 2`] = `
Object {
  "limit": 100,
  "node1_Nid": "Vid",
}
`;

exports[`one long-branch with value leaf WHERE (NOT exists (path-to-long-branch-with-value)) 1`] = `"MATCH (root:\`Root\`) WHERE (NOT exists((root)-[:\`LINKED_TO1\`]->(:\`Node\`)-[:\`LINKED_TO2\`]->(:\`Node\`{Nid:$node2_Nid}))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one long-branch with value leaf WHERE (NOT exists (path-to-long-branch-with-value)) 2`] = `
Object {
  "limit": 100,
  "node2_Nid": "Vid",
}
`;

exports[`one long-branch with value mid WHERE (NOT exists (path-to-long-branch-with-value)) 1`] = `"MATCH (root:\`Root\`), (root:\`Root\`)-[r0:\`LINKED_TO1\`]->(node1:\`Node\`) WHERE node1.Nid = $node1_Nid AND (NOT exists((root)-[:\`LINKED_TO1\`]->(:\`Node\`)-[:\`LINKED_TO2\`]->(:\`Node\`))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one long-branch with value mid WHERE (NOT exists (path-to-long-branch-with-value)) 2`] = `
Object {
  "limit": 100,
  "node1_Nid": "Vid",
}
`;

exports[`one long-branch with value mid and leaf WHERE (NOT exists (path-to-long-branch-with-value)) 1`] = `"MATCH (root:\`Root\`), (root:\`Root\`)-[r0:\`LINKED_TO1\`]->(node1:\`Node\`) WHERE node1.Nid = $node1_Nid AND (NOT exists((root)-[:\`LINKED_TO1\`]->(:\`Node\`)-[:\`LINKED_TO2\`]->(:\`Node\`{Nid:$node2_Nid}))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one long-branch with value mid and leaf WHERE (NOT exists (path-to-long-branch-with-value)) 2`] = `
Object {
  "limit": 100,
  "node1_Nid": "Vid",
  "node2_Nid": "Vid2",
}
`;

exports[`one very long-branch with value leaf and reverse order generation 1`] = `"MATCH (root:\`Root\`) WHERE (NOT exists((root)-[:\`LINKED_TO1\`]->(:\`Label1\`)<-[:\`LINKED_TO2\`]-(:\`Label2\`)-[:\`LINKED_TO3\`]->(:\`Label3\`{Nid:$label3_Nid}))) RETURN DISTINCT root.name AS name, root.id AS id LIMIT $limit"`;

exports[`one very long-branch with value leaf and reverse order generation 2`] = `
Object {
  "label3_Nid": "Vid3",
  "limit": 100,
}
`;
